File: examples/python/01_coin_flip.py
"""
01 - Coin Flip (Beta-Binomial Model)

The simplest Bayesian model: estimate the probability of heads from
observed coin flips. We place a Beta(1,1) prior (uniform on [0,1])
on theta and observe 65 heads out of 100 flips.

Analytic posterior: Beta(66, 36), mean = 66/102 ~ 0.647
"""

import bayesiangpu as bg

# --- Data ---
n_trials = 100
n_heads = 65

# --- Model ---
model = bg.Model()
model.param("theta", bg.Beta(1, 1))          # Prior: uniform on [0,1]
model.observe(bg.Binomial(n_trials, "theta"), [float(n_heads)])

# --- Inference ---
result = bg.sample(model, num_samples=1000, num_chains=4, seed=42)

# --- Results ---
s = result.summarize("theta")
analytic_mean = (1 + n_heads) / (2 + n_trials)

print("=" * 50)
print("Coin Flip: Beta-Binomial Model")
print("=" * 50)
print(f"Data: {n_heads} heads in {n_trials} flips")
print()
print(f"Posterior mean:  {s.mean:.4f}  (analytic: {analytic_mean:.4f})")
print(f"Posterior std:   {s.std:.4f}")
print(f"95% CI:          [{s.q025:.4f}, {s.q975:.4f}]")
print(f"R-hat:           {s.rhat:.4f}")
print(f"ESS:             {s.ess:.0f}")
print()
print(f"Converged: {result.is_converged()}")

File: examples/r/01_coin_flip.R
# 01_coin_flip.R -- Beta-Binomial coin flip
#
# The simplest Bayesian model: estimate the bias of a coin.
# Prior:      theta ~ Beta(1, 1)  (uniform on [0,1])
# Likelihood: data  ~ Binomial(100, theta)
# Observed:   65 heads out of 100 flips
#
# Analytical posterior: Beta(1 + 65, 1 + 35) = Beta(66, 36)
# Posterior mean: 66/102 ~ 0.647

library(bayesiangpu)

# --- Build model (method chaining) ----------------------------------------
model <- Model$new()
model$param("theta", Beta(1, 1))
model$observe(Binomial(100, "theta"), 65)

cat("=== Coin Flip Model ===\n")
print(model)

# --- Sample -----------------------------------------------------------------
result <- bg_sample(model, num_samples = 2000, num_chains = 4, seed = 42)
print(result)

# --- Compare to analytical posterior ----------------------------------------
s <- result$summarize("theta")
analytical_mean <- 66 / 102

cat("\n--- Comparison ---\n")
cat(sprintf("Estimated mean:  %.4f\n", s$mean))
cat(sprintf("Analytical mean: %.4f\n", analytical_mean))
cat(sprintf("95%% CI:          [%.4f, %.4f]\n", s$q025, s$q975))
cat(sprintf("R-hat: %.4f  ESS: %.0f\n", s$rhat, s$ess))
